pkg/status/combinedstatus-resolver.go

```diff
@@ -0,0 +1,472 @@
+	// destinations set to collect from, shared for all workload objects
+	collectionDestinations := sets.New(abstract.SliceMap(bindingResolution.Destinations,
+		func(destination v1alpha1.Destination) string { return destination.ClusterId })...)
+
+	// (1)
+	if deleted {
+		// for every combinedstatus resolution - queue its object identifier for syncing
+		// and remove the resolution from memory
+		for workloadObjectIdentifier, _ := range c.bindingNameToResolutions[bindingName] {
+			combinedStatusIdentifiersToQueue.Insert(getCombinedStatusIdentifier(bindingName, workloadObjectIdentifier))
+		}
+
+		delete(c.bindingNameToResolutions, bindingName)
+		return combinedStatusIdentifiersToQueue
+	}
+
+	// if the binding resolution is not yet noted - create a new entry
+	objectIdentifierToResolution, exists := c.bindingNameToResolutions[bindingName]
+	if !exists {
+		objectIdentifierToResolution = make(map[util.ObjectIdentifier]*combinedStatusResolution)
+		c.bindingNameToResolutions[bindingName] = objectIdentifierToResolution
+	}
+
+	// (2) remove excessive combinedstatus resolutions of objects that are no longer
+	// associated with the binding resolution
+	for objectIdentifier, _ := range objectIdentifierToResolution {
+		if _, exists := bindingResolution.ObjectIdentifierToData[objectIdentifier]; !exists {
+			combinedStatusIdentifiersToQueue.Insert(getCombinedStatusIdentifier(bindingName, objectIdentifier))
+			delete(objectIdentifierToResolution, objectIdentifier)
+		}
+	}
+
+	// (~2+3) create/update combinedstatus resolutions for every object that requires status collection,
+	// and delete resolutions that are no longer required
+	for objectIdentifier, objectData := range bindingResolution.ObjectIdentifierToData {
+		csResolution, exists := objectIdentifierToResolution[objectIdentifier]
+		if len(objectData.StatusCollectors) == 0 {
+			if exists { // associated resolution is no longer required
+				combinedStatusIdentifiersToQueue.Insert(getCombinedStatusIdentifier(bindingName, objectIdentifier))
+				delete(objectIdentifierToResolution, objectIdentifier)
+			}
+
+			continue
+		}
+
+		// create resolution entry if missing
+		if !exists {
+			csResolution = &combinedStatusResolution{
+				statusCollectorNameToData: make(map[string]*statusCollectorData),
+			}
+			objectIdentifierToResolution[objectIdentifier] = csResolution
+		}
+
+		// fetch missing statuscollector specs
+		c.fetchMissingStatusCollectorSpecsLocked(statusCollectorLister, objectData.StatusCollectors)
+
+		// update statuscollectors
+		removedCollectors, addedCollectors := csResolution.setStatusCollectors(
+			abstract.SliceToPrimitiveMap(objectData.StatusCollectors,
+				func(statusCollectorName string) string { return statusCollectorName }, // keys mapper
+				func(statusCollectorName string) *v1alpha1.StatusCollectorSpec { // val mapper
+					return c.statusCollectorNameToSpec[statusCollectorName] // if missing after fetch, ignored
```

Comment:
More precisely, the value returned here will be `nil` if the earlier fetch failed on this statusCollectorName. I suspect that `csResolution.setStatusCollectors` will do the wrong thing for the `map` entry created in this case.

Reply:
`csResolution.setStatusCollectors` would create an entry for the status collector with a nil spec. While this would not do anything harmful (as it would be expected for ::updateStatusCollector to be called with the same name at a certain stage) - it would mark the combined-status resolution as updated, which is inaccurate. Fixing the latter.